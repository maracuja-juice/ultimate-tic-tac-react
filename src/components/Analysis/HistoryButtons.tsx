import * as React from 'react';
import { Button } from '@rmwc/button';
import { scroller } from 'react-scroll/modules';
import { moveScrollElementBaseName } from './ScrollElementConstants';

interface HistoryButtonsProps {
  currentMove: number;
  lastMove: number;
  moveForwardInHistory: (numberOfMoves: number) => void;
  moveBackwardInHistory: (numberOfMoves: number) => void;
}

export class HistoryButtons extends React.Component<HistoryButtonsProps> {
  handleEvent = (event: KeyboardEvent) => {
    const {
      currentMove,
      lastMove,
      moveBackwardInHistory,
      moveForwardInHistory,
    } = this.props;

    if (
      (event.key === 'ArrowLeft' || event.key === 'Left') &&
      this.props.currentMove !== 1
    ) {
      moveBackwardInHistory(1);
      this.scrollToElement(currentMove - 1);
    }

    if (
      (event.key === 'ArrowRight' || event.key === 'Right') &&
      currentMove !== lastMove
    ) {
      moveForwardInHistory(1);
      this.scrollToElement(currentMove + 1);
    }
  };

  scrollToElement = (moveNumberToScrollTo: number) => {
    scroller.scrollTo(moveScrollElementBaseName + moveNumberToScrollTo, {
      duration: 300,
      smooth: true,
      containerId: 'moveList',
      offset: -108,
    });
  };

  componentDidMount(): void {
    window.addEventListener('keydown', this.handleEvent);
  }

  componentWillUnmount(): void {
    window.removeEventListener('keydown', this.handleEvent);
  }

  render() {
    const {
      currentMove,
      lastMove,
      moveForwardInHistory,
      moveBackwardInHistory,
    } = this.props;

    return (
      <div>
        <Button
          disabled={currentMove === 1}
          dense={true}
          raised={true}
          icon="arrow-left"
          onClick={() => {
            moveBackwardInHistory(1);
            this.scrollToElement(currentMove - 1);
          }}
        >
          Previous
        </Button>
        <Button
          disabled={currentMove === lastMove}
          dense={true}
          raised={true}
          trailingIcon="arrow-right"
          onClick={() => {
            moveForwardInHistory(1);
            this.scrollToElement(currentMove + 1);
          }}
        >
          Next
        </Button>
      </div>
    );
  }
}
