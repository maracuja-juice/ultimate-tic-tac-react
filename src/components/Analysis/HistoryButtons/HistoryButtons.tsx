import * as React from 'react';
import { Button } from '@rmwc/button';
import styles from './HistoryButtons.module.css';

interface HistoryButtonsProps {
  currentMove: number;
  lastMove: number;
  moveForwardInHistory: (numberOfMoves: number) => void;
  moveBackwardInHistory: (numberOfMoves: number) => void;
  onInteraction: (newMoveNumber: number) => void;
}

export class HistoryButtons extends React.Component<HistoryButtonsProps> {
  handleEvent = (event: KeyboardEvent) => {
    const {
      currentMove,
      lastMove,
      moveBackwardInHistory,
      moveForwardInHistory,
    } = this.props;

    if (
      (event.key === 'ArrowLeft' || event.key === 'Left') &&
      this.props.currentMove !== 1
    ) {
      moveBackwardInHistory(1);
      this.props.onInteraction(currentMove - 1);
    }

    if (
      (event.key === 'ArrowRight' || event.key === 'Right') &&
      currentMove !== lastMove
    ) {
      moveForwardInHistory(1);
      this.props.onInteraction(currentMove + 1);
    }
  };

  componentDidMount(): void {
    window.addEventListener('keydown', this.handleEvent);
  }

  componentWillUnmount(): void {
    window.removeEventListener('keydown', this.handleEvent);
  }

  render() {
    const {
      currentMove,
      lastMove,
      moveForwardInHistory,
      moveBackwardInHistory,
      onInteraction,
    } = this.props;

    return (
      <div>
        <Button
          disabled={currentMove === 1}
          dense={true}
          raised={true}
          icon="arrow-left"
          onClick={() => {
            moveBackwardInHistory(1);
            onInteraction(currentMove - 1);
          }}
          className={styles.buttonMargin}
        >
          Previous
        </Button>
        <Button
          disabled={currentMove === lastMove}
          dense={true}
          raised={true}
          trailingIcon="arrow-right"
          onClick={() => {
            moveForwardInHistory(1);
            onInteraction(currentMove + 1);
          }}
          className={styles.buttonMargin}
        >
          Next
        </Button>
      </div>
    );
  }
}
