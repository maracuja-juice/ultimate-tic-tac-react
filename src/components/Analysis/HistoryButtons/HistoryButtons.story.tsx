import * as React from 'react';
import { Story, Meta } from '@storybook/react/types-6-0';
import { HistoryButtons, HistoryButtonsProps } from './HistoryButtons';
import { action } from '@storybook/addon-actions';
import { IconProvider } from '../../IconProvider';

export default {
  title: 'HistoryButtons',
  component: HistoryButtons,
} as Meta;

export const Default: Story<HistoryButtonsProps> = (args) => (
  <IconProvider>
    <div className="historyButtons">
      <HistoryButtons
        {...args}
        moveForwardInHistory={action('moveForwardInHistory')}
        moveBackwardInHistory={action('moveBackwardInHistory')}
        onInteraction={action('interaction')}
      />
    </div>
  </IconProvider>
);
Default.args = {
  currentMove: 1,
  lastMove: 5,
};
