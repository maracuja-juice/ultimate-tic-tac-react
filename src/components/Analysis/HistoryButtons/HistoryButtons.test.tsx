import * as React from 'react';
import { shallow } from 'enzyme';
import { HistoryButtons } from './HistoryButtons';
import { Button } from '@rmwc/button';

describe('HistoryButtons', function () {
  it('should match snapshot', () => {
    const moveForwardInHistory = jest.fn((numberOfMoves) => {});
    const moveBackwardInHistory = jest.fn((numberOfMoves) => {});
    const onInteraction = jest.fn((newMoveNumber) => {});

    const historyButtons = shallow(
      <HistoryButtons
        currentMove={1}
        lastMove={15}
        moveForwardInHistory={moveForwardInHistory}
        moveBackwardInHistory={moveBackwardInHistory}
        onInteraction={onInteraction}
      />,
    );

    expect(historyButtons).toMatchSnapshot();
  });

  describe('disabled behaviour', () => {
    it('has a disabled move backward button when current move is 1', () => {
      const moveForwardInHistory = jest.fn((numberOfMoves) => {});
      const moveBackwardInHistory = jest.fn((numberOfMoves) => {});
      const onInteraction = jest.fn((newMoveNumber) => {});

      const historyButtons = shallow(
        <HistoryButtons
          currentMove={1}
          lastMove={15}
          moveForwardInHistory={moveForwardInHistory}
          moveBackwardInHistory={moveBackwardInHistory}
          onInteraction={onInteraction}
        />,
      );

      const backwardButton = historyButtons.find(Button).get(0);
      const forwardButton = historyButtons.find(Button).get(1);

      expect(backwardButton.props.disabled).toBe(true);
      expect(forwardButton.props.disabled).toBe(false);
    });

    it('has a disabled move forward button when current move is the last move', () => {
      const moveForwardInHistory = jest.fn((numberOfMoves) => {});
      const moveBackwardInHistory = jest.fn((numberOfMoves) => {});
      const onInteraction = jest.fn((newMoveNumber) => {});

      const historyButtons = shallow(
        <HistoryButtons
          currentMove={15}
          lastMove={15}
          moveForwardInHistory={moveForwardInHistory}
          moveBackwardInHistory={moveBackwardInHistory}
          onInteraction={onInteraction}
        />,
      );

      const backwardButton = historyButtons.find(Button).get(0);
      const forwardButton = historyButtons.find(Button).get(1);

      expect(backwardButton.props.disabled).toBe(false);
      expect(forwardButton.props.disabled).toBe(true);
    });
  });

  describe('click behaviour', () => {
    it('calls forward when move forward button is clicked', () => {
      const moveForwardInHistory = jest.fn((numberOfMoves) => {});
      const moveBackwardInHistory = jest.fn((numberOfMoves) => {});
      const onInteraction = jest.fn((newMoveNumber) => {});

      const historyButtons = shallow(
        <HistoryButtons
          currentMove={1}
          lastMove={15}
          moveForwardInHistory={moveForwardInHistory}
          moveBackwardInHistory={moveBackwardInHistory}
          onInteraction={onInteraction}
        />,
      );

      const forwardButton = historyButtons.find(Button).at(1);
      forwardButton.simulate('click');

      expect(moveForwardInHistory).toHaveBeenCalledWith(1);
    });

    it('calls backward when move backward button is clicked', () => {
      const moveForwardInHistory = jest.fn((numberOfMoves) => {});
      const moveBackwardInHistory = jest.fn((numberOfMoves) => {});
      const onInteraction = jest.fn((newMoveNumber) => {});

      const historyButtons = shallow(
        <HistoryButtons
          currentMove={1}
          lastMove={15}
          moveForwardInHistory={moveForwardInHistory}
          moveBackwardInHistory={moveBackwardInHistory}
          onInteraction={onInteraction}
        />,
      );

      const backwardButton = historyButtons.find(Button).first();
      backwardButton.simulate('click');

      expect(moveBackwardInHistory).toHaveBeenCalledWith(1);
    });
  });

  describe('keyboard event behaviour', () => {
    it('calls backward when left arrow is pressed', () => {
      const moveForwardInHistory = jest.fn((numberOfMoves) => {});
      const moveBackwardInHistory = jest.fn((numberOfMoves) => {});
      const onInteraction = jest.fn((newMoveNumber) => {});

      shallow(
        <HistoryButtons
          currentMove={5}
          lastMove={15}
          moveForwardInHistory={moveForwardInHistory}
          moveBackwardInHistory={moveBackwardInHistory}
          onInteraction={onInteraction}
        />,
      );

      const event = new KeyboardEvent('keydown', { key: 'ArrowLeft' });
      window.dispatchEvent(event);

      expect(moveBackwardInHistory).toHaveBeenCalledWith(1);
      expect(moveForwardInHistory).toHaveBeenCalledTimes(0);
    });

    it('calls forward when right arrow is pressed', () => {
      const moveForwardInHistory = jest.fn((numberOfMoves) => {});
      const moveBackwardInHistory = jest.fn((numberOfMoves) => {});
      const onInteraction = jest.fn((newMoveNumber) => {});

      shallow(
        <HistoryButtons
          currentMove={5}
          lastMove={15}
          moveForwardInHistory={moveForwardInHistory}
          moveBackwardInHistory={moveBackwardInHistory}
          onInteraction={onInteraction}
        />,
      );

      const event = new KeyboardEvent('keydown', { key: 'ArrowRight' });
      window.dispatchEvent(event);

      expect(moveForwardInHistory).toHaveBeenCalledWith(1);
      expect(moveBackwardInHistory).toHaveBeenCalledTimes(0);
    });

    it('doesnt call backward when left arrow is pressed and moveNumber is 1', () => {
      const moveForwardInHistory = jest.fn((numberOfMoves) => {});
      const moveBackwardInHistory = jest.fn((numberOfMoves) => {});
      const onInteraction = jest.fn((newMoveNumber) => {});

      shallow(
        <HistoryButtons
          currentMove={1}
          lastMove={15}
          moveForwardInHistory={moveForwardInHistory}
          moveBackwardInHistory={moveBackwardInHistory}
          onInteraction={onInteraction}
        />,
      );

      const event = new KeyboardEvent('keydown', { key: 'ArrowLeft' });
      window.dispatchEvent(event);

      expect(moveBackwardInHistory).toHaveBeenCalledTimes(0);
      expect(moveForwardInHistory).toHaveBeenCalledTimes(0);
    });

    it('doesnt call forward when right arrow is pressed and moveNumber is the last one', () => {
      const moveForwardInHistory = jest.fn((numberOfMoves) => {});
      const moveBackwardInHistory = jest.fn((numberOfMoves) => {});
      const onInteraction = jest.fn((newMoveNumber) => {});

      shallow(
        <HistoryButtons
          currentMove={15}
          lastMove={15}
          moveForwardInHistory={moveForwardInHistory}
          moveBackwardInHistory={moveBackwardInHistory}
          onInteraction={onInteraction}
        />,
      );

      const event = new KeyboardEvent('keydown', { key: 'ArrowRight' });
      window.dispatchEvent(event);

      expect(moveBackwardInHistory).toHaveBeenCalledTimes(0);
      expect(moveForwardInHistory).toHaveBeenCalledTimes(0);
    });
  });
});
