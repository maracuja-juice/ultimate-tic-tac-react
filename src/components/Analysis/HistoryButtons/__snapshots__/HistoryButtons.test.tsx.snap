// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistoryButtons should match snapshot 1`] = `
<Fragment>
  <withRipple(Button)
    className="buttonMargin"
    dense={true}
    disabled={true}
    icon="arrow-left"
    onClick={[Function]}
    raised={true}
  >
    Previous
  </withRipple(Button)>
  <withRipple(Button)
    className="buttonMargin"
    dense={true}
    disabled={false}
    onClick={[Function]}
    raised={true}
    trailingIcon="arrow-right"
  >
    Next
  </withRipple(Button)>
</Fragment>
`;
